sys/touch_dev: add generic API for touch device

aabadie opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

This PR introduces a new generic API for handling touch device like the stmpe811 one.

For the moment, only single touch position is handled but maybe the current design could support multi touch devices.

The API could maybe be extended to support gestures (see #13214).

I'm planning to open a follow-up PR that extends the lvgl integration with this API.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

There is a test application provided that should work as is on `stm32f429i-disc1`:

<details><summary>make -C tests/touch_dev flash term --no-print-directory</summary>

```
Building application "tests_touch_dev" for "stm32f429i-disc1" with MCU "stm32f4".

"make" -C /work/riot/RIOT/boards/stm32f429i-disc1
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/stm32f4
"make" -C /work/riot/RIOT/cpu/cortexm_common
"make" -C /work/riot/RIOT/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT/cpu/stm32_common
"make" -C /work/riot/RIOT/cpu/stm32_common/periph
"make" -C /work/riot/RIOT/cpu/stm32f4/periph
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/drivers/stmpe811
"make" -C /work/riot/RIOT/drivers/touch_dev
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/auto_init
"make" -C /work/riot/RIOT/sys/div
"make" -C /work/riot/RIOT/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT/sys/pm_layered
"make" -C /work/riot/RIOT/sys/stdio_uart
"make" -C /work/riot/RIOT/sys/xtimer
   text	   data	    bss	    dec	    hex	filename
  12896	    112	   2748	  15756	   3d8c	/work/riot/RIOT/tests/touch_dev/bin/stm32f429i-disc1/tests_touch_dev.elf
/work/riot/RIOT/dist/tools/openocd/openocd.sh flash /work/riot/RIOT/tests/touch_dev/bin/stm32f429i-disc1/tests_touch_dev.elf
### Flashing Target ###
Open On-Chip Debugger 0.10.0+dev-01047-g09ac9ab1 (2020-02-05-08:53)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
WARNING: interface/stlink-v2-1.cfg is deprecated, please switch to interface/stlink.cfg
hla_swd
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : clock speed 2000 kHz
Info : STLINK V2J25M14 (API v2) VID:PID 0483:374B
Info : Target voltage: 2.852356
Error: jtag status contains invalid mode value - communication failure
Polling target stm32f4x.cpu failed, trying to reexamine
Examination failed, GDB will be halted. Polling again in 100ms
Info : Previous state query failed, trying to reconnect
Error: jtag status contains invalid mode value - communication failure
Polling target stm32f4x.cpu failed, trying to reexamine
Examination failed, GDB will be halted. Polling again in 300ms
Info : Listening on port 35245 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f4x.cpu       hla_target little stm32f4x.cpu       unknown

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Error: mem2array: Read @ 0xe0042004, w=4, cnt=1, failed
Error executing event examine-end on target stm32f4x.cpu:
/usr/local/bin/../share/openocd/scripts/mem_helper.tcl:6: Error: 
in procedure 'ocd_process_reset' 
in procedure 'ocd_process_reset_inner' called at file "embedded:startup.tcl", line 270
in procedure 'mmw' called at file "/usr/local/bin/../share/openocd/scripts/target/stm32f4x.cfg", line 79
in procedure 'mrw' called at file "/usr/local/bin/../share/openocd/scripts/mem_helper.tcl", line 36
at file "/usr/local/bin/../share/openocd/scripts/mem_helper.tcl", line 6
Info : Previous state query failed, trying to reconnect
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x08000524 msp: 0x20000200
Polling target stm32f4x.cpu failed, trying to reexamine
Info : stm32f4x.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : device id = 0x20016419
Info : flash size = 2048 kbytes
Info : Dual Bank 2048 kiB STM32F42x/43x/469/479 found
auto erase enabled
wrote 16384 bytes from file /work/riot/RIOT/tests/touch_dev/bin/stm32f429i-disc1/tests_touch_dev.elf in 0.677064s (23.631 KiB/s)

verified 13008 bytes in 0.165338s (76.831 KiB/s)

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
shutdown command invoked
Done flashing
/work/riot/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" 
2020-04-16 10:33:31,122 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
2020-04-16 10:33:54,058 # Pressed!
2020-04-16 10:33:54,059 # X: 99, Y:160
2020-04-16 10:33:54,076 # X: 98, Y:161
2020-04-16 10:33:54,088 # X: 99, Y:161
2020-04-16 10:33:54,100 # X: 100, Y:160
2020-04-16 10:33:54,112 # X: 100, Y:160
2020-04-16 10:33:54,124 # Released!
2020-04-16 10:33:54,166 # Pressed!
2020-04-16 10:33:54,172 # X: 98, Y:156
2020-04-16 10:33:54,184 # X: 98, Y:156
2020-04-16 10:33:54,196 # X: 98, Y:156
2020-04-16 10:33:54,208 # X: 98, Y:157
2020-04-16 10:33:54,226 # X: 98, Y:157
2020-04-16 10:33:54,238 # X: 98, Y:156
2020-04-16 10:33:54,250 # X: 98, Y:156
2020-04-16 10:33:54,262 # X: 98, Y:155
2020-04-16 10:33:54,274 # Pressed!
2020-04-16 10:33:54,275 # X: 97, Y:153
2020-04-16 10:33:54,286 # Released!
2020-04-16 10:33:57,244 # Pressed!
2020-04-16 10:33:57,250 # X: 140, Y:153
2020-04-16 10:33:57,268 # X: 140, Y:154
2020-04-16 10:33:57,274 # X: 140, Y:153
2020-04-16 10:33:57,286 # X: 141, Y:153
2020-04-16 10:33:57,303 # X: 141, Y:153
2020-04-16 10:33:57,316 # X: 142, Y:153
2020-04-16 10:33:57,328 # X: 142, Y:153
2020-04-16 10:33:57,340 # X: 143, Y:153
2020-04-16 10:33:57,352 # X: 144, Y:153
2020-04-16 10:33:57,369 # X: 144, Y:152
2020-04-16 10:33:57,382 # X: 145, Y:152
2020-04-16 10:33:57,394 # X: 145, Y:152
2020-04-16 10:33:57,406 # X: 145, Y:152
2020-04-16 10:33:57,418 # X: 145, Y:152
2020-04-16 10:33:57,435 # X: 145, Y:152
2020-04-16 10:33:57,448 # X: 146, Y:152
2020-04-16 10:33:57,460 # X: 147, Y:152
2020-04-16 10:33:57,472 # X: 147, Y:152
2020-04-16 10:33:57,484 # X: 147, Y:152
```

</details>

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

None

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

kaspar030 commented:
Should we think of multi-touch already? What is prev_x/y needed for?

kaspar030 commented:
> Should we think of multi-touch already?

reading helps

aabadie commented:
> What is prev_x/y needed for?

This is part of the stmpe811 driver and it was added to dismiss invalid coordinates. It appears in the diff because the alignment of the doxygen comments was shifted a bit.

leandrolanzieri commented on drivers/include/touch_dev.h:
```diff
@@ -0,0 +1,132 @@
+ */
+typedef struct {
+    uint16_t x;     /**< X coordinate */
+    uint16_t y;     /**< X coordinate */
```
```suggestion
    uint16_t y;     /**< Y coordinate */
```

leandrolanzieri commented on drivers/include/touch_dev.h:
```diff
@@ -0,0 +1,132 @@
+     * @brief   Get the current pressed position of the touch device
+     *
+     * @param[in] dev       Pointer to the touch device
+     * @param[out] position The position pressed
```
What about also returning `bool` to indicate if it is being pressed or not and the position in one function call?

aabadie replied:
I was looking at #13214 and maybe we should return the number of touch points and 0 means no touch. We could indeed remove the is_pressed function from the API.
What do you think ?

leandrolanzieri replied:
Would then make sense to pass an array of `touch_position_t` here? meaning that we should also indicate the length of it. I would remove `is_pressed` if we allow here to pass `NULL` pointer to just ask for the number of touches.

aabadie replied:
Ok, let's do this then.

leandrolanzieri commented on drivers/include/touch_dev.h:
```diff
@@ -0,0 +1,132 @@
+ *
+ * @return              Height in points
+ */
+uint16_t touch_dev_height(touch_dev_t *dev);
```
Could the functions be `const touch_dev_t *dev`?

aabadie replied:
I don't think that makes a bug difference, but that's a good idea indeed. Let's do this then.

aabadie commented:
@leandrolanzieri, I addressed your comments. The new version is 36B slower on ROM and still works:

<details>

```
make -C tests/touch_dev/ flash term
make: Entering directory '/work/riot/RIOT/tests/touch_dev'
Building application "tests_touch_dev" for "stm32f429i-disc1" with MCU "stm32f4".

"make" -C /work/riot/RIOT/boards/stm32f429i-disc1
"make" -C /work/riot/RIOT/core
"make" -C /work/riot/RIOT/cpu/stm32f4
"make" -C /work/riot/RIOT/cpu/cortexm_common
"make" -C /work/riot/RIOT/cpu/cortexm_common/periph
"make" -C /work/riot/RIOT/cpu/stm32_common
"make" -C /work/riot/RIOT/cpu/stm32_common/periph
"make" -C /work/riot/RIOT/cpu/stm32f4/periph
"make" -C /work/riot/RIOT/drivers
"make" -C /work/riot/RIOT/drivers/periph_common
"make" -C /work/riot/RIOT/drivers/stmpe811
"make" -C /work/riot/RIOT/drivers/touch_dev
"make" -C /work/riot/RIOT/sys
"make" -C /work/riot/RIOT/sys/auto_init
"make" -C /work/riot/RIOT/sys/div
"make" -C /work/riot/RIOT/sys/newlib_syscalls_default
"make" -C /work/riot/RIOT/sys/pm_layered
"make" -C /work/riot/RIOT/sys/stdio_uart
"make" -C /work/riot/RIOT/sys/xtimer
   text	   data	    bss	    dec	    hex	filename
  12860	    112	   2748	  15720	   3d68	/work/riot/RIOT/tests/touch_dev/bin/stm32f429i-disc1/tests_touch_dev.elf
/work/riot/RIOT/dist/tools/openocd/openocd.sh flash /work/riot/RIOT/tests/touch_dev/bin/stm32f429i-disc1/tests_touch_dev.elf
### Flashing Target ###
Open On-Chip Debugger 0.10.0+dev-01047-g09ac9ab1 (2020-02-05-08:53)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
WARNING: interface/stlink-v2-1.cfg is deprecated, please switch to interface/stlink.cfg
hla_swd
Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
Info : clock speed 2000 kHz
Info : STLINK V2J25M14 (API v2) VID:PID 0483:374B
Info : Target voltage: 2.852356
Error: jtag status contains invalid mode value - communication failure
Polling target stm32f4x.cpu failed, trying to reexamine
Examination failed, GDB will be halted. Polling again in 100ms
Info : Previous state query failed, trying to reconnect
Error: jtag status contains invalid mode value - communication failure
Polling target stm32f4x.cpu failed, trying to reexamine
Examination failed, GDB will be halted. Polling again in 300ms
Info : Listening on port 36199 for gdb connections
    TargetName         Type       Endian TapName            State       
--  ------------------ ---------- ------ ------------------ ------------
 0* stm32f4x.cpu       hla_target little stm32f4x.cpu       unknown

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Error: mem2array: Read @ 0xe0042004, w=4, cnt=1, failed
Error executing event examine-end on target stm32f4x.cpu:
/usr/local/bin/../share/openocd/scripts/mem_helper.tcl:6: Error: 
in procedure 'ocd_process_reset' 
in procedure 'ocd_process_reset_inner' called at file "embedded:startup.tcl", line 270
in procedure 'mmw' called at file "/usr/local/bin/../share/openocd/scripts/target/stm32f4x.cfg", line 79
in procedure 'mrw' called at file "/usr/local/bin/../share/openocd/scripts/mem_helper.tcl", line 36
at file "/usr/local/bin/../share/openocd/scripts/mem_helper.tcl", line 6
Info : Previous state query failed, trying to reconnect
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x080004f4 msp: 0x20000200
Polling target stm32f4x.cpu failed, trying to reexamine
Info : stm32f4x.cpu: hardware has 6 breakpoints, 4 watchpoints
Info : device id = 0x20016419
Info : flash size = 2048 kbytes
Info : Dual Bank 2048 kiB STM32F42x/43x/469/479 found
auto erase enabled
wrote 16384 bytes from file /work/riot/RIOT/tests/touch_dev/bin/stm32f429i-disc1/tests_touch_dev.elf in 0.688017s (23.255 KiB/s)

verified 12972 bytes in 0.168534s (75.166 KiB/s)

Info : Unable to match requested speed 2000 kHz, using 1800 kHz
Info : Unable to match requested speed 2000 kHz, using 1800 kHz
shutdown command invoked
Done flashing
/work/riot/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" 
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2020-04-28 16:18:33,525 # Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
2020-04-28 16:18:37,048 # Pressed!
2020-04-28 16:18:37,049 # X: 152, Y:195
2020-04-28 16:18:37,061 # X: 151, Y:194
2020-04-28 16:18:37,073 # X: 151, Y:195
2020-04-28 16:18:37,090 # X: 151, Y:195
2020-04-28 16:18:37,103 # X: 151, Y:195
2020-04-28 16:18:37,114 # X: 152, Y:195
2020-04-28 16:18:37,127 # X: 152, Y:195
2020-04-28 16:18:37,139 # X: 152, Y:195
2020-04-28 16:18:37,156 # X: 152, Y:195
2020-04-28 16:18:37,168 # X: 152, Y:195
2020-04-28 16:18:37,180 # X: 152, Y:195
2020-04-28 16:18:37,192 # X: 152, Y:195
2020-04-28 16:18:37,204 # X: 153, Y:195
2020-04-28 16:18:37,222 # X: 153, Y:195
2020-04-28 16:18:37,235 # X: 153, Y:195
2020-04-28 16:18:37,247 # X: 152, Y:195
2020-04-28 16:18:37,259 # X: 152, Y:195
2020-04-28 16:18:37,276 # X: 152, Y:194
2020-04-28 16:18:37,288 # X: 152, Y:194
2020-04-28 16:18:37,301 # X: 152, Y:194
2020-04-28 16:18:37,313 # X: 152, Y:195
2020-04-28 16:18:37,325 # X: 152, Y:195
2020-04-28 16:18:37,337 # X: 152, Y:195
2020-04-28 16:18:37,354 # X: 151, Y:195
2020-04-28 16:18:37,367 # X: 150, Y:195
2020-04-28 16:18:37,379 # X: 149, Y:196
2020-04-28 16:18:37,391 # X: 149, Y:196
2020-04-28 16:18:37,403 # X: 149, Y:196
2020-04-28 16:18:37,420 # X: 149, Y:196
2020-04-28 16:18:37,433 # X: 149, Y:196
2020-04-28 16:18:37,445 # X: 149, Y:196
2020-04-28 16:18:37,457 # X: 150, Y:196
2020-04-28 16:18:37,468 # Pressed!
2020-04-28 16:18:37,469 # X: 153, Y:195
2020-04-28 16:18:37,486 # Released!
2020-04-28 16:18:40,997 # Pressed!
2020-04-28 16:18:41,003 # X: 83, Y:252
2020-04-28 16:18:41,014 # X: 84, Y:250
2020-04-28 16:18:41,026 # X: 87, Y:249
2020-04-28 16:18:41,038 # X: 88, Y:250
2020-04-28 16:18:41,050 # X: 88, Y:250
2020-04-28 16:18:41,068 # X: 88, Y:251
2020-04-28 16:18:41,081 # X: 88, Y:251
2020-04-28 16:18:41,092 # X: 88, Y:251
```

</details>

leandrolanzieri commented on drivers/stmpe811/stmpe811_touch_dev.c:
```diff
@@ -0,0 +1,72 @@
+    assert(dev);
+
+    /* stmpe811 is only single touch */
+    assert(len == 1);
```
Is this check really needed? If an application uses this interface to have device-agnostic access to the touch sensor it might pass always a pre-allocated array.

aabadie replied:
Ok, then removed !

aabadie commented:
> Please address and squash directly

Done!
